File: core-scala/src/main/scala/io/asterisque/wire/gateway/netty/WebSocket.scala
package io.asterisque.wire.gateway.netty

import java.net.{InetSocketAddress, URI}
import java.util.Objects

import io.asterisque.core.wire.netty.HTTP
import io.asterisque.utils.Debug
import io.netty.bootstrap.{Bootstrap, ServerBootstrap}
import io.netty.buffer.ByteBufUtil
import io.netty.channel._
import io.netty.channel.socket.nio.{NioServerSocketChannel, NioSocketChannel}
import io.netty.handler.codec.http._
import io.netty.handler.codec.http.websocketx.WebSocketClientProtocolHandler.ClientHandshakeStateEvent
import io.netty.handler.codec.http.websocketx._
import io.netty.handler.ssl.{SslContext, SslContextBuilder, SslHandler}
import io.netty.handler.stream.ChunkedWriteHandler
import javax.annotation.{Nonnull, Nullable}
import org.slf4j.LoggerFactory

import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.{Future, Promise}
import scala.util.{Failure, Success}

/**
  * Netty を使用した WebSocket 通信のためのライブラリ。
  */
object WebSocket {
  private[this] val SUBPROTOCOL = "v10.asterisque"

  /**
    * WebSocket プロトコルを使用するサーバクラス。
    */
  object Server {
    private[Server] val logger = LoggerFactory.getLogger(classOf[WebSocket.Server])

    trait Listener {
      def wsServerReady(@Nonnull ch:Channel):Unit = {
        logger.debug(s"wsServerReady($ch)")
      }

      /**
        * @param ctx コンテキスト。bind に失敗した場合などコンテキストが確定する前のエラーでは null
        * @param ex  発生した例外
        */
      def wsServerCaughtException(@Nullable ctx:ChannelHandlerContext, @Nonnull ex:Throwable):Unit = {
        logger.debug(s"wsServerCaughtException($ctx, $ex)")
      }
    }

    /**
      * WebSocket サーバが接続を受け付けたときにチャネルの初期化を行います。
      */
    private[Server] class Initializer(val subprotocol:String, val uri:URI, val sslContext:SslContext, val onAccept:Channel => WebSocket.Servant) extends ChannelInitializer[Channel] {
      override protected def initChannel(@Nonnull ch:Channel):Unit = {
        logger.trace("initChannel({})", ch)

        val path = Option(uri.getPath).filter(_.nonEmpty).getOrElse("/")
        val subprotocol = SUBPROTOCOL + "," + this.subprotocol
        if(validAndSecureScheme(uri)) {
          val engine = sslContext.newEngine(ch.alloc)
          ch.pipeline.addFirst("tls", new SslHandler(engine))
        }
        ch.pipeline
          .addLast("http", new HttpServerCodec)
          .addLast("http-chunked", new ChunkedWriteHandler)
          .addLast("aggregator", new HttpObjectAggregator(64 * 1024))
          .addLast("request", new WebSocket.HttpRequestHandler(path))
          .addLast("websocket", new WebSocketServerProtocolHandler(path, subprotocol))
          .addLast("io.asterisque.server", new WebSocket.WebSocketFrameHandler(onAccept(ch)))
      }
    }

  }

  /**
    * WebSocket プロトコルを使用するサーバクラス。
    *
    * @param group       サーバで使用するイベントループ
    * @param subprotocol asterisque 上で使用するサブプロトコル
    * @param listener    サーバイベントリスナ
    * @param sslContext  SSL/TLS を使用する場合は有効な { @link SslContext}、使用しない場合は null
    */
  class Server(val group:EventLoopGroup, val subprotocol:String, val listener:Server.Listener, val sslContext:SslContext) {
    private[this] val logger = Server.logger

    private var channel:Channel = _

    /**
      * 指定されたソケットアドレスにサーバソケットを bind します。bind に成功した場合 [[Server.Listener.wsServerReady()]]
      * コールバックが発生します。
      * bind に失敗した場合は例外の Future を返します ([[Server.Listener]] へのコールバックではありません)。
      *
      * @param uri      バインドアドレスとパスの URI 表現 (e.g., "wss://0.0.0.0:9999/ws")。
      * @param onAccept accept ごとに呼び出されチャネルから Servant を生成する
      * @return サーバソケットのチャネル
      */
    def bind(@Nonnull uri:URI, @Nonnull onAccept:Channel => WebSocket.Servant):Future[Channel] = {
      logger.trace("bind({})", uri)

      val port = math.max(0, uri.getPort)
      val host = Option(uri.getHost).getOrElse("localhost")
      val address = new InetSocketAddress(host, port)

      if(sslContext == null && validAndSecureScheme(uri)) {
        throw new IllegalArgumentException(s"secure WebSocket specified but ssl context is null: $uri")
      }

      val bootstrap = new ServerBootstrap()
        .group(group)
        .channel(classOf[NioServerSocketChannel])
        .handler(new ChannelInitializer[Channel]() {
          override protected def initChannel(ch:Channel):Unit = {
            listener.wsServerReady(ch)
            // super.initChannel(ch) is an abstract method
          }

          override def exceptionCaught(ctx:ChannelHandlerContext, cause:Throwable):Unit = {
            listener.wsServerCaughtException(ctx, cause)
            super.exceptionCaught(ctx, cause)
          }
        })
        .childHandler(new Server.Initializer(subprotocol, uri, sslContext, onAccept))

      bootstrap.bind(address).toFuture.andThen {
        case Success(ch) =>
          channel = ch
          logger.debug("the server has completed binding: {}", channel)
        case Failure(ex) =>
          listener.wsServerCaughtException(null, ex)
      }
    }

    def destroy():Unit = {
      logger.trace("destroy()")
      if(channel != null) {
        channel.close()
      }
    }
  }

  /**
    * WebSocket プロトコルを使用するクライアントクラス。
    */
  object Client {
    private val logger = LoggerFactory.getLogger(classOf[WebSocket.Client])

    private[this] val DEFAULT_SSL_CONTEXT = SslContextBuilder.forClient().build()

    /**
      * WebSocket サーバと接続したときにチャネルの初期化を行います。
      */
    private[Client] class Initializer(val uri:URI, val subprotocol:String, val servant:Channel => Servant, val sslContext:SslContext) extends ChannelInitializer[Channel] {
      override protected def initChannel(@Nonnull ch:Channel):Unit = {
        logger.trace("initChannel({})", ch)

        val subprotocol = s"$SUBPROTOCOL, ${this.subprotocol}"
        if(validAndSecureScheme(uri)) {
          val engine = (if(sslContext != null) sslContext else DEFAULT_SSL_CONTEXT).newEngine(ch.alloc)
          ch.pipeline.addFirst("tls", new SslHandler(engine))
        }
        ch.pipeline
          .addLast("http", new HttpClientCodec)
          .addLast("http-chunked", new ChunkedWriteHandler)
          .addLast("aggregator", new HttpObjectAggregator(64 * 1024))
          .addLast("websocket", new WebSocketClientProtocolHandler(uri, WebSocketVersion.V13, subprotocol, false, EmptyHttpHeaders.INSTANCE, 64 * 1024))
          .addLast("io.asterisque.client", new WebSocket.WebSocketFrameHandler(servant(ch)))
      }
    }

  }

  /**
    * @param group       クライアントで使用するイベントループ
    * @param subprotocol asterisque 上で使用するサブプロトコル
    * @param sslContext  SSL/TLS を使用する場合は有効な { @link SslContext}、使用しない場合は null
    */
  class Client(@Nonnull val group:EventLoopGroup, @Nonnull val subprotocol:String, @Nullable val sslContext:SslContext) {
    private[this] val logger = Client.logger
    Objects.requireNonNull(group)
    Objects.requireNonNull(subprotocol)

    private var channel:Channel = _

    def connect(@Nonnull uri:URI, @Nonnull servant:Channel => Servant):Future[Channel] = {
      Objects.requireNonNull(uri)
      Objects.requireNonNull(servant)
      logger.trace("connect({})", uri)
      new Bootstrap()
        .group(group)
        .channel(classOf[NioSocketChannel])
        .handler(new Client.Initializer(uri, subprotocol, servant, sslContext))
        .connect(new InetSocketAddress(uri.getHost, uri.getPort))
        .toFuture.andThen {
        case Success(ch) =>
          channel = ch
          logger.debug("the client has completed binding: {}", channel)
        case Failure(_) => ()
      }
    }

    def destroy():Unit = {
      Client.logger.trace("destroy()")
      if(channel != null) {
        channel.close()
      }
    }
  }

  /**
    * WebSocket の双方のエンドポイントでフレームの送受信を行うためのインターフェース。
    */
  trait Servant {

    /**
      * WebSocket ハンドシェイクが完了し WebSocket のフレームが送受信可能になったときに呼び出されます。
      *
      * @param ctx WebSocket プロトコルの準備が完了したコンテキスト
      */
    def wsReady(@Nonnull ctx:ChannelHandlerContext):Unit

    /**
      * 相手側から WebSocket のフレームを受信したときに呼び出されます。
      *
      * @param ctx フレームを受信したコンテキスト
      * @param msg 受信したフレーム
      */
    def wsFrameReceived(@Nonnull ctx:ChannelHandlerContext, @Nonnull msg:WebSocketFrame):Unit

    /**
      * 指定されたチャネルがクローズされているときに呼び出されます。
      *
      * @param ctx クローズされているチャネル
      */
    def wsClosed(@Nonnull ctx:ChannelHandlerContext):Unit

    /**
      * 指定されたコンテキスト上で例外が発生したときに呼び出されます。
      *
      * @param ctx 例外の発生したコンテキスト
      * @param ex  発生した例外
      */
    def wsCaughtException(@Nonnull ctx:ChannelHandlerContext, @Nonnull ex:Throwable):Unit
  }

  /**
    * WebSocket のフレームを受信します。
    * WS ハンドシェイクが完了したときに HttpRequestHandler をパイプラインから取り除く。
    */
  private[this] object WebSocketFrameHandler {
    private val logger = LoggerFactory.getLogger(classOf[WebSocket.WebSocketFrameHandler])
  }

  private class WebSocketFrameHandler(val servant:WebSocket.Servant) extends SimpleChannelInboundHandler[WebSocketFrame] {
    private[this] val logger = WebSocketFrameHandler.logger
    private var handshakeState:ClientHandshakeStateEvent = _

    @throws[Exception]
    override def channelActive(ctx:ChannelHandlerContext):Unit = {
      logger.trace("channelActive({})", ctx)
      super.channelActive(ctx)
    }

    @throws[Exception]
    override def channelInactive(ctx:ChannelHandlerContext):Unit = {
      logger.trace("channelInactive({})", ctx)
      if(handshakeState == ClientHandshakeStateEvent.HANDSHAKE_ISSUED) {
        servant.wsCaughtException(ctx, new WebSocket.HandshakeException("WebSocket handshake failure"))
      }
      servant.wsClosed(ctx)
      super.channelInactive(ctx)
    }

    @throws[Exception]
    override def exceptionCaught(ctx:ChannelHandlerContext, cause:Throwable):Unit = {
      logger.trace(s"exceptionCaught($ctx, $cause)")
      servant.wsCaughtException(ctx, cause)
      super.exceptionCaught(ctx, cause)
    }

    @throws[Exception]
    override def userEventTriggered(@Nonnull ctx:ChannelHandlerContext, @Nonnull evt:Any):Unit = {
      logger.trace("userEventTriggered({},{})", ctx, evt)
      evt match {
        case _:WebSocketServerProtocolHandler.HandshakeComplete =>
          logger.trace("server handshake complete")
          handshakeState = ClientHandshakeStateEvent.HANDSHAKE_COMPLETE
          ctx.pipeline.remove(classOf[WebSocket.HttpRequestHandler])
          servant.wsReady(ctx)
        case ClientHandshakeStateEvent.HANDSHAKE_COMPLETE =>
          logger.trace("client handshake complete")
          handshakeState = ClientHandshakeStateEvent.HANDSHAKE_COMPLETE
          servant.wsReady(ctx)
        case ClientHandshakeStateEvent.HANDSHAKE_ISSUED =>
          // HANDSHAKE_ISSUED の後に HANDSHAKE_COMPLETE が来るためこの時点ではエラーではない
          logger.trace("client handshake uncompleted")
          handshakeState = WebSocketClientProtocolHandler.ClientHandshakeStateEvent.HANDSHAKE_ISSUED
        case _ =>
          super.userEventTriggered(ctx, evt)
      }
    }

    override protected def channelRead0(@Nonnull ctx:ChannelHandlerContext, @Nonnull msg:WebSocketFrame):Unit = {
      logger.trace(s"channelRead0($ctx, $msg)")
      if(logger.isTraceEnabled) {
        msg match {
          case frame:TextWebSocketFrame =>
            logger.trace("RECEIVE: {}", Debug.toString(frame.text))
          case _ =>
            val buf = msg.content
            logger.trace("RECEIVE: {}", Debug.toString(ByteBufUtil.getBytes(buf)))
        }
      }
      servant.wsFrameReceived(ctx, msg)
    }
  }

  /**
    * WebSocket ハンドシェイクの前段階で WebSocket 用の HTTP リクエストパスのみを許可するハンドラ。ハンドシェイク後は
    * パイプラインから取り除かれる。
    */
  private class HttpRequestHandler(val path:String) extends SimpleChannelInboundHandler[FullHttpRequest] {
    private[this] val logger = LoggerFactory.getLogger(classOf[HttpRequestHandler])

    override def channelRead0(@Nonnull ctx:ChannelHandlerContext, @Nonnull request:FullHttpRequest):Unit = {
      if(path.equalsIgnoreCase(request.uri)) {
        ctx.fireChannelRead(request.retain)
      } else {
        val msg = s"asterisque WebSocket bridge doesn't work for the specified URI: ${request.uri}"
        ctx.writeAndFlush(HTTP.newErrorResponse(HttpResponseStatus.FORBIDDEN, msg))
          .addListener { _:io.netty.util.concurrent.Future[_ >: Void] => ctx.disconnect() }
        logger.debug(msg)
      }
    }
  }

  /**
    * Netty の ChannelFuture を CompletableFuture に変換する。
    *
    * @param cf ChannelFuture
    */
  private[this] implicit class _ChannelFuture(cf:ChannelFuture) {
    def toFuture:Future[Channel] = {
      val promise = Promise[Channel]()
      cf.addListener { _:io.netty.util.concurrent.Future[_] =>
        if(cf.isSuccess) {
          promise.success(cf.channel)
        } else {
          promise.failure(cf.cause)
        }
      }
      promise.future
    }
  }

  /**
    * 指定された URI スキームがセキュア WebSocket かアンセキュア WebSocket かを判定します。
    *
    * @param uri セキュアを判定する URI
    * @throws IllegalArgumentException URI スキームを認識できない場合
    * @return セキュア WebSocket の場合 true
    */
  @throws[IllegalArgumentException]
  private[this] def validAndSecureScheme(uri:URI):Boolean = {
    Option(uri.getScheme).map(_.toLowerCase).getOrElse("wss") match {
      case "ws" => false
      case "wss" => true
      case scheme =>
        throw new IllegalArgumentException(s"unsupported WebSocket scheme '$scheme': $uri")
    }
  }

  /**
    * WebSocket のハンドシェイクに失敗したときに発生する例外。
    */
  class HandshakeException(@Nonnull val msg:String) extends RuntimeException(msg)

}

File: core-scala/src/main/java/io/asterisque/core/wire/netty/WebSocket.java
package io.asterisque.core.wire.netty;

import io.asterisque.utils.Debug;
import io.netty.bootstrap.Bootstrap;
import io.netty.bootstrap.ServerBootstrap;
import io.netty.buffer.ByteBuf;
import io.netty.buffer.ByteBufUtil;
import io.netty.channel.*;
import io.netty.channel.socket.nio.NioServerSocketChannel;
import io.netty.channel.socket.nio.NioSocketChannel;
import io.netty.handler.codec.http.*;
import io.netty.handler.codec.http.websocketx.*;
import io.netty.handler.ssl.SslContext;
import io.netty.handler.ssl.SslHandler;
import io.netty.handler.stream.ChunkedWriteHandler;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import javax.net.ssl.SSLEngine;
import java.net.InetSocketAddress;
import java.net.SocketAddress;
import java.net.URI;
import java.util.concurrent.CompletableFuture;
import java.util.function.Function;

public class WebSocket {

  private static final String SUBPROTOCOL = "v10.asterisque";

  private WebSocket() {
  }

  /**
   * WebSocket プロトコルを使用するサーバクラス。
   */
  public static class Server {
    private static final Logger logger = LoggerFactory.getLogger(Server.class);

    public static final Listener EmptyListener = new Listener() {
      @Override
      public void wsServerReady(@Nonnull Channel ch) {
        logger.trace("wsServerReady({})", ch);
      }

      @Override
      public void wsServerCaughtException(@Nullable ChannelHandlerContext ctx, @Nonnull Throwable ex) {
        logger.trace("wsServerCaughtException({},{})", ctx, ex);
      }
    };

    private final EventLoopGroup group;
    private Channel channel;

    private final String subprotocol;
    private final String path;
    private final Server.Listener listener;
    private final SslContext sslContext;

    /**
     * WebSocket プロトコルを使用するサーバクラス。
     *
     * @param group       サーバで使用するイベントループ
     * @param subprotocol asterisque 上で使用するサブプロトコル
     * @param path        WebSocket が使用する HTTP URI パス
     * @param listener    サーバイベントリスナ
     * @param sslContext  SSL/TLS を使用する場合は有効な {@link SslContext}、使用しない場合は null
     */
    public Server(@Nonnull EventLoopGroup group,
                  @Nonnull String subprotocol, @Nonnull String path, @Nonnull Server.Listener listener,
                  @Nullable SslContext sslContext) {
      if (!path.startsWith("/")) {
        throw new IllegalArgumentException("the server path must begin with '/': '" + path + "'");
      }
      this.group = group;
      this.subprotocol = subprotocol;
      this.path = path;
      this.listener = listener;
      this.sslContext = sslContext;
    }

    /**
     * SSL/TLS を使用せず WebSocket プロトコルを使用するサーバクラス。
     *
     * @param group       サーバで使用するイベントループ
     * @param subprotocol asterisque 上で使用するサブプロトコル
     * @param path        WebSocket が使用する HTTP URI パス
     */
    public Server(@Nonnull EventLoopGroup group, @Nonnull String subprotocol, @Nonnull String path) {
      this(group, subprotocol, path, EmptyListener, null);
    }

    /**
     * 指定されたソケットアドレスにサーバソケットを bind します。bind に成功した場合 {@link Listener#wsServerReady(Channel)}
     * コールバックが発生します。
     * bind に失敗した場合は例外の Future を返します ({@link Listener} へのコールバックではありません)。
     *
     * @param address  バインドアドレス
     * @param onAccept accept ごとに呼び出されチャネルから Servant を生成する
     * @return サーバソケットのチャネル
     */
    public CompletableFuture<Channel> bind(@Nonnull SocketAddress address,
                                           @Nonnull Function<Channel, Servant> onAccept) {
      logger.trace("bind({})", address);

      ServerBootstrap bootstrap = new ServerBootstrap();
      bootstrap.group(group)
          .channel(NioServerSocketChannel.class)
          .handler(new ChannelInitializer<Channel>() {
            @Override
            protected void initChannel(Channel ch) {
              listener.wsServerReady(ch);
            }

            @Override
            public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
              listener.wsServerCaughtException(ctx, cause);
              super.exceptionCaught(ctx, cause);
            }
          })
          .childHandler(new Initializer(subprotocol, path, sslContext, onAccept));

      return channelFutureToFuture(bootstrap.bind(address)).whenComplete((ch, ex) -> {
        if (ex == null) {
          channel = ch;
          logger.debug("the server has completed binding: {}", channel);
        } else {
          listener.wsServerCaughtException(null, ex);
        }
      });
    }

    public void destroy() {
      logger.trace("destroy()");
      if (channel != null) {
        channel.close();
      }
    }

    public interface Listener {
      void wsServerReady(@Nonnull Channel ch);

      void wsServerCaughtException(@Nullable ChannelHandlerContext ctx, @Nonnull Throwable ex);
    }

    /**
     * WebSocket サーバが接続を受け付けたときにチャネルの初期化を行います。
     */
    private static class Initializer extends ChannelInitializer<Channel> {
      private static final Logger logger = LoggerFactory.getLogger(Initializer.class);

      private final String subprotocol;
      private final String path;
      private final SslContext sslContext;
      private final Function<Channel, Servant> onAccept;

      private Initializer(@Nonnull String subprotocol, @Nonnull String path, @Nullable SslContext sslContext,
                          @Nonnull Function<Channel, Servant> onAccept) {
        this.subprotocol = subprotocol;
        this.path = path;
        this.sslContext = sslContext;
        this.onAccept = onAccept;
      }

      @Override
      protected void initChannel(@Nonnull Channel ch) {
        logger.trace("initChannel({})", ch);
        Servant servant = onAccept.apply(ch);
        String subprotocol = SUBPROTOCOL + "," + this.subprotocol;
        if (sslContext != null) {
          SSLEngine engine = sslContext.newEngine(ch.alloc());
          ch.pipeline().addFirst("tls", new SslHandler(engine));
        }
        ch.pipeline()
            .addLast("http", new HttpServerCodec())
            .addLast("http-chunked", new ChunkedWriteHandler())
            .addLast("aggregator", new HttpObjectAggregator(64 * 1024))
            .addLast("request", new HttpRequestHandler(path))
            .addLast("websocket", new WebSocketServerProtocolHandler(path, subprotocol))
            .addLast("io.asterisque.server", new WebSocketFrameHandler(servant));
      }
    }
  }

  /**
   * WebSocket プロトコルを使用するクライアントクラス。
   */
  public static class Client {
    private static final Logger logger = LoggerFactory.getLogger(Client.class);

    private final EventLoopGroup group;
    private Channel channel;

    private final String subprotocol;
    private final Servant servant;
    private final SslContext sslContext;

    /**
     * @param group       クライアントで使用するイベントループ
     * @param subprotocol asterisque 上で使用するサブプロトコル
     * @param servant     WebSocket イベントごとのコールバック
     * @param sslContext  SSL/TLS を使用する場合は有効な {@link SslContext}、使用しない場合は null
     */
    public Client(@Nonnull EventLoopGroup group, @Nonnull String subprotocol, @Nonnull Servant servant,
                  @Nullable SslContext sslContext) {
      this.group = group;
      this.subprotocol = subprotocol;
      this.servant = servant;
      this.sslContext = sslContext;
    }

    public CompletableFuture<Channel> connect(@Nonnull URI uri) {
      logger.trace("connect({})", uri);

      Bootstrap bootstrap = new Bootstrap().group(group).channel(NioSocketChannel.class)
          .handler(new Initializer(uri, subprotocol, servant, sslContext));

      return channelFutureToFuture(bootstrap.connect(new InetSocketAddress(uri.getHost(), uri.getPort())))
          .whenComplete((ch, ex) -> {
            if (ex != null) {
              channel = ch;
            }
          });
    }

    public void destroy() {
      logger.trace("destroy()");
      if (channel != null) {
        channel.close();
      }
    }

    /**
     * WebSocket サーバと接続したときにチャネルの初期化を行います。
     */
    private static class Initializer extends ChannelInitializer<Channel> {
      private static final Logger logger = LoggerFactory.getLogger(Initializer.class);

      private final URI uri;
      private final String subprotocol;
      private final Servant servant;
      private final SslContext sslContext;

      private Initializer(@Nonnull URI uri, @Nonnull String subprotocol, @Nonnull Servant servant,
                          @Nullable SslContext sslContext) {
        this.uri = uri;
        this.subprotocol = subprotocol;
        this.servant = servant;
        this.sslContext = sslContext;
      }

      @Override
      protected void initChannel(@Nonnull Channel ch) {
        logger.trace("initChannel({})", ch);
        String subprotocol = SUBPROTOCOL + ", " + this.subprotocol;
        if (sslContext != null) {
          SSLEngine engine = sslContext.newEngine(ch.alloc());
          ch.pipeline().addFirst("tls", new SslHandler(engine));
        }
        ch.pipeline()
            .addLast("http", new HttpClientCodec())
            .addLast("http-chunked", new ChunkedWriteHandler())
            .addLast("aggregator", new HttpObjectAggregator(64 * 1024))
            .addLast("websocket", new WebSocketClientProtocolHandler(uri, WebSocketVersion.V13, subprotocol,
                false, EmptyHttpHeaders.INSTANCE, 64 * 1024))
            .addLast("io.asterisque.client", new WebSocketFrameHandler(servant));
      }
    }
  }

  /**
   * WebSocket の双方のエンドポイントでフレームの送受信を行うためのインターフェース。
   */
  public interface Servant {
    /**
     * WebSocket ハンドシェイクが完了し WebSocket のフレームが送受信可能になったときに呼び出されます。
     *
     * @param ctx WebSocket プロトコルの準備が完了したコンテキスト
     */
    void wsReady(@Nonnull ChannelHandlerContext ctx);

    /**
     * 相手側から WebSocket のフレームを受信したときに呼び出されます。
     *
     * @param ctx フレームを受信したコンテキスト
     * @param msg 受信したフレーム
     */
    void wsFrameReceived(@Nonnull ChannelHandlerContext ctx, @Nonnull WebSocketFrame msg);

    /**
     * 指定されたチャネルがクローズされているときに呼び出されます。
     *
     * @param ctx クローズされているチャネル
     */
    void wsClosed(@Nonnull ChannelHandlerContext ctx);

    /**
     * 指定されたコンテキスト上で例外が発生したときに呼び出されます。
     *
     * @param ctx 例外の発生したコンテキスト
     * @param ex  発生した例外
     */
    void wsCaughtException(@Nonnull ChannelHandlerContext ctx, @Nonnull Throwable ex);
  }

  /**
   * WebSocket のフレームを受信します。
   * WS ハンドシェイクが完了したときに HttpRequestHandler をパイプラインから取り除く。
   */
  private static class WebSocketFrameHandler extends SimpleChannelInboundHandler<WebSocketFrame> {
    private static final Logger logger = LoggerFactory.getLogger(WebSocketFrameHandler.class);

    private final Servant servant;

    private WebSocketClientProtocolHandler.ClientHandshakeStateEvent handshakeState;

    private WebSocketFrameHandler(@Nonnull Servant servant) {
      this.servant = servant;
    }

    @Override
    public void channelActive(ChannelHandlerContext ctx) throws Exception {
      logger.trace("channelActive({})", ctx);
      super.channelActive(ctx);
    }

    @Override
    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
      logger.trace("channelInactive({})", ctx);
      if (handshakeState == WebSocketClientProtocolHandler.ClientHandshakeStateEvent.HANDSHAKE_ISSUED) {
        servant.wsCaughtException(ctx, new HandshakeException("WebSocket handshake failure"));
      }
      servant.wsClosed(ctx);
      super.channelInactive(ctx);
    }

    @Override
    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
      logger.trace("exceptionCaught({},{})", ctx, cause);
      servant.wsCaughtException(ctx, cause);
      super.exceptionCaught(ctx, cause);
    }

    @Override
    public void userEventTriggered(@Nonnull ChannelHandlerContext ctx, @Nonnull Object evt) throws Exception {
      logger.trace("userEventTriggered({},{})", ctx, evt);
      if (evt instanceof WebSocketServerProtocolHandler.HandshakeComplete) {
        logger.trace("server handshake complete");
        handshakeState = WebSocketClientProtocolHandler.ClientHandshakeStateEvent.HANDSHAKE_COMPLETE;
        ctx.pipeline().remove(HttpRequestHandler.class);
        servant.wsReady(ctx);
      } else if (evt == WebSocketClientProtocolHandler.ClientHandshakeStateEvent.HANDSHAKE_COMPLETE) {
        logger.trace("client handshake complete");
        handshakeState = WebSocketClientProtocolHandler.ClientHandshakeStateEvent.HANDSHAKE_COMPLETE;
        servant.wsReady(ctx);
      } else if (evt == WebSocketClientProtocolHandler.ClientHandshakeStateEvent.HANDSHAKE_ISSUED) {
        logger.trace("client handshake uncompleted");
        handshakeState = WebSocketClientProtocolHandler.ClientHandshakeStateEvent.HANDSHAKE_ISSUED;
        // HANDSHAKE_ISSUED の後に HANDSHAKE_COMPLETE が来るためこの時点ではエラーではない
      } else {
        super.userEventTriggered(ctx, evt);
      }
    }

    @Override
    protected void channelRead0(@Nonnull ChannelHandlerContext ctx, @Nonnull WebSocketFrame msg) {
      logger.trace("channelRead0({},{})", ctx, msg);
      if (logger.isTraceEnabled()) {
        if (msg instanceof TextWebSocketFrame) {
          logger.trace("RECEIVE: {}", Debug.toString(((TextWebSocketFrame) msg).text()));
        } else {
          ByteBuf buf = msg.content();
          logger.trace("RECEIVE: {}", Debug.toString(ByteBufUtil.getBytes(buf)));
        }
      }

      servant.wsFrameReceived(ctx, msg);
    }
  }

  /**
   * WebSocket ハンドシェイクの前段階で WebSocket 用の HTTP リクエストパスのみを許可するハンドラ。ハンドシェイク後は
   * パイプラインから取り除かれる。
   */
  private static class HttpRequestHandler extends SimpleChannelInboundHandler<FullHttpRequest> {
    private final String path;

    private HttpRequestHandler(@Nonnull String path) {
      this.path = path;
    }

    @Override
    public void channelRead0(@Nonnull ChannelHandlerContext ctx, @Nonnull FullHttpRequest request) {
      if (path.equalsIgnoreCase(request.uri())) {
        ctx.fireChannelRead(request.retain());
      } else {
        String msg = "asterisque WebSocket bridge doesn't work for the specified URI: " + request.uri();
        ctx.writeAndFlush(HTTP.newErrorResponse(HttpResponseStatus.FORBIDDEN, msg))
            .addListener(future -> ctx.disconnect());
      }
    }
  }

  /**
   * Netty の ChannelFuture を CompletableFuture に変換する。
   *
   * @param cf ChannelFuture
   * @return CompletableFuture
   */
  @Nonnull
  private static CompletableFuture<Channel> channelFutureToFuture(@Nonnull ChannelFuture cf) {
    CompletableFuture<Channel> future = new CompletableFuture<>();
    cf.addListener(ignore -> {
      if (cf.isSuccess()) {
        future.complete(cf.channel());
      } else {
        future.completeExceptionally(cf.cause());
      }
    });
    return future;
  }

  /**
   * WebSocket のハンドシェイクに失敗したときに発生する例外。
   */
  public static class HandshakeException extends RuntimeException {
    public HandshakeException(@Nonnull String msg) {
      super(msg);
    }
  }
}
